### ISSUE TITLE
Alterações no código

### ISSUE THREAD

**Issue by VascoLopo** [VascoLopo/ajuda#6]
import spacy
from mlxtend.plotting import plot_decision_regions
from sentence_transformers import SentenceTransformer
from collections import defaultdict
import json
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import itertools
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score
import numpy as np
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix
import matplotlib.pyplot as plt
# Load the data
with open(r'C:\Users\Lopo\Downloads\project-10-at-2023-04-19-17-02-b9209ecb.json', encoding='utf-8') as f:
    data = json.load(f)
# Load the Portuguese model
nlp = spacy.load("pt_core_news_lg")
# Define the stopwords set
wordz = {"de", "a", "o", "que", "e", "do", "ao", "rt", "da", "em", "um", "para", "é", "com", "uma",
             "os", "no", "se", "na", "por", "mais", "as", "dos", "como", "mas", "foi", "ao", "ele", "das",
             "tem", "à", "seu", "sua", "ou", "ser", "quando", "muito", "há", "nos", "já", "está", "eu",
             "também", "só", "pelo", "pela", "até", "isso", "ela", "entre", "era", "depois", "sem", "mesmo",
             "aos", "ter", "seus", "quem", "nas", "me", "esse", "eles", "estão", "você", "tinha", "foram",
             "essa", "num", "nem", "suas", "meu", "às", "minha", "têm", "numa", "pelos", "elas", "havia",
             "seja", "qual", "será", "nós", "tenho", "lhe", "deles", "essas", "esses", "pelas", "este", "fosse",
             "dele", "tu", "te", "vocês", "vos", "hei", "há", "havemos", "hão", "houve", "houvemos", "houveram",
             "houvera", "houvéramos", "haja", "hajamos", "hajam", "for", "is", "are", "this", "the", "and", "with",
             "to", "into", "on", "in", "of", "at", " . ", " , "}

# Define a SentenceTransformer model
#model = SentenceTransformer('sentence-transformers/paraphrase-xlm-r-multilingual-v1')
model = SentenceTransformer('xlm-r-100langs-bert-base-nli-mean-tokens')
embeddings = [[] for _ in range(len(data))]
labels = [[] for _ in range(len(data))]
pos_tags = [[] for _ in range(len(data))]
parse_trees = [[] for _ in range(len(data))]

for i, item in enumerate(data):
    id = i
    annotations = item['annotations']
    doc = nlp(item['data']['text'])
    tokens = [token.lemma_.lower() for token in doc if token.text.lower() not in wordz]
    preprocessed_sentence = " ".join(tokens)
    for annotation in annotations:
        result = annotation['result']
        if result:
            value = result[0]['value']
            choices = value.get('choices', [])
            if choices:
                choice = choices[0]
                embedding = model.encode(preprocessed_sentence)
                pos_tag = [[token.text, token.pos_] for token in doc]
                parse_tree = str(list(doc.sents)[0].root)
                embeddings[id].append(embedding)
                labels[id].append(choice)
                pos_tags[id].append(pos_tag)
                parse_trees[id].append(parse_tree)

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    print("ID:", i)
    print("Embeddings:", embeddings[i])
    print("Labels:", labels[i])
    print("POS tags:", pos_tags[i])
    print("Parse trees:", parse_trees[i])
    print("")
import pandas as pd

# Create an empty DataFrame with the desired columns
df = pd.DataFrame(columns=['ID', 'Embeddings', 'Labels', 'POS Tags', 'Parse Trees'])

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    # Modify the 'labels' list so that any case where labels are both 'Populism' and 'Non populism' is counted only as 'Populism'
    if 'Non populism' in labels[i] and 'Populism' not in labels[i]:
        labels[i] = ['Non populism']
    elif 'Populism' in labels[i]:
        labels[i] = ['Populism']

    # Create a new row of data for each item
    row_data = {'ID': i, 'Embeddings': embeddings[i], 'Labels': labels[i], 'POS Tags': pos_tags[i],
                'Parse Trees': parse_trees[i]}
    # Append the row to the DataFrame
    df = df.append(row_data, ignore_index=True)

# Set the display option to show all columns
pd.set_option('display.max_columns', None)

# Print the DataFrame
print(df.head(10))

sizes = set()
for i in range(len(embeddings)):
    for embedding in embeddings[i]:
        size = len(embedding)
        sizes.add(size)
print(sizes)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(embeddings, labels, test_size=0.25, random_state=42)

# Reshape the data to 2D arrays
X_train = [np.concatenate(x) for x in X_train]
X_test = [np.concatenate(x) for x in X_test]
y_train = np.array([y[0] for y in y_train])
y_test = [y[0] for y in y_test]
# Fit a logistic regression model to the training data
clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)
# Define the SVM with an RBF kernel
svm = SVC(kernel='rbf', random_state=42)

# Define the hyperparameters to search over
param_grid = {'C': [0.1, 1, 10], 'gamma': [0.1, 1, 'scale']}

# Use GridSearchCV to search over the hyperparameters
grid_search = GridSearchCV(svm, param_grid, cv=5, verbose=2)

# Train the SVM using the best hyperparameters found by GridSearchCV
grid_search.fit(list(X_train), list(y_train))

# Predict on the test set and print the accuracy
predictions = grid_search.predict(list(X_train))
accuracy = accuracy_score(list(y_train), predictions)
print("Accuracy:", accuracy)
#---------------------------------------------------------------------testing
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.colors import ListedColormap
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import make_pipeline
from sklearn.datasets import make_moons, make_circles, make_classification
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
from sklearn.inspection import DecisionBoundaryDisplay

names = [
    "Nearest Neighbors",
    "Linear SVM",
    "RBF SVM",
    "Gaussian Process",
    "Decision Tree",
    "Random Forest",
    "Neural Net",
    "AdaBoost",
    "Naive Bayes",
    "QDA",
]

classifiers = [
    KNeighborsClassifier(3),
    SVC(kernel="linear", C=0.025),
    SVC(gamma=2, C=1),
    GaussianProcessClassifier(1.0 * RBF(1.0)),
    DecisionTreeClassifier(max_depth=5),
    RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1),
    MLPClassifier(alpha=1, max_iter=1000),
    AdaBoostClassifier(),
    GaussianNB(),
    QuadraticDiscriminantAnalysis(),
]

X, y = make_classification(
    n_features=2, n_redundant=0, n_informative=2, random_state=1, n_clusters_per_class=1
)
rng = np.random.RandomState(2)
X += 2 * rng.uniform(size=X.shape)
linearly_separable = (X, y)

datasets = [
    make_moons(noise=0.3, random_state=0),
    make_circles(noise=0.2, factor=0.5, random_state=1),
    linearly_separable,
]

figure = plt.figure(figsize=(27, 9))
i = 1
# iterate over datasets
for ds_cnt, ds in enumerate(datasets):
    # preprocess dataset, split into training and test part
    X, y = ds
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.4, random_state=42
    )

    x_min, x_max = X[:, 0].min() - 0.5, X[:, 0].max() + 0.5
    y_min, y_max = X[:, 1].min() - 0.5, X[:, 1].max() + 0.5

    # just plot the dataset first
    cm = plt.cm.RdBu
    cm_bright = ListedColormap(["#FF0000", "#0000FF"])
    ax = plt.subplot(len(datasets), len(classifiers) + 1, i)
    if ds_cnt == 0:
        ax.set_title("Input data")
    # Plot the training points
    ax.scatter(X_train[:, 0], X_train[:, 1], c=y_train, cmap=cm_bright, edgecolors="k")
    # Plot the testing points
    ax.scatter(
        X_test[:, 0], X_test[:, 1], c=y_test, cmap=cm_bright, alpha=0.6, edgecolors="k"
    )
    ax.set_xlim(x_min, x_max)
    ax.set_ylim(y_min, y_max)
    ax.set_xticks(())
    ax.set_yticks(())
    i += 1

    # iterate over classifiers
    for name, clf in zip(names, classifiers):
        ax = plt.subplot(len(datasets), len(classifiers) + 1, i)

        clf = make_pipeline(StandardScaler(), clf)
        clf.fit(X_train, y_train)
        score = clf.score(X_test, y_test)
        DecisionBoundaryDisplay.from_estimator(
            clf, X, cmap=cm, alpha=0.8, ax=ax, eps=0.5
        )

        # Plot the training points
        ax.scatter(
            X_train[:, 0], X_train[:, 1], c=y_train, cmap=cm_bright, edgecolors="k"
        )
        # Plot the testing points
        ax.scatter(
            X_test[:, 0],
            X_test[:, 1],
            c=y_test,
            cmap=cm_bright,
            edgecolors="k",
            alpha=0.6,
        )

        ax.set_xlim(x_min, x_max)
        ax.set_ylim(y_min, y_max)
        ax.set_xticks(())
        ax.set_yticks(())
        if ds_cnt == 0:
            ax.set_title(name)
        ax.text(
            x_max - 0.3,
            y_min + 0.3,
            ("%.2f" % score).lstrip("0"),
            size=15,
            horizontalalignment="right",
        )
        i += 1
# Label the red and blue dots
for j, label in enumerate(["red", "blue"]):
    ax.text(
        x_min + 0.5, y_min + 0.5 + j * 0.1,
        f"{label} dots = {np.sum(y_train==j)} (train) / {np.sum(y_test==j)} (test)",
        size=10,
        color=label,
        bbox=dict(facecolor='w', edgecolor=label, boxstyle='round,pad=0.3'),
        horizontalalignment='left',
        verticalalignment='bottom',
        transform=ax.transAxes,
    )
plt.tight_layout()
plt.show()
